### ISSUE TITLE
rpath needed in criterion.pc (v2.4-rc)

### ISSUE THREAD

**Issue by kugel-** [Snaipe/Criterion#415]
Hello,

criterion.pc does not add `-Wl,-rpath=${libdir}` to `Libs:`. When I install criterion to a non-default location then I can link but not execute my tests, because libcriterion.so is not found at runtime.

I can't remember that this was a problem in 2.3.3 but I didn't double check that yet.

**Comment by MrAnno** [comment]
Hi,

Thank you for opening this issue.

This was also a problem in v2.3.3.

I'm not 100% sure we should solve this on the pkg-config level. Setting RPATH unconditionally is undesirable, especially by distro maintainers.

When using pkg-config together with libtool, the issue is fixed automatically because libtool adds the necessary flags when it detects non-system libraries.

**Comment by kugel-** [comment]
Well, normally this would be sufficient

```
diff --git a/src/criterion.pc.in b/src/criterion.pc.in
index 23737a4..9a6b01e 100644
--- a/src/criterion.pc.in
+++ b/src/criterion.pc.in
@@ -8,5 +8,5 @@ Description: A KISS, Cross platform unit testing framework for C and C++
 URL: https://github.com/Snaipe/Criterion/
 Version: @PROJECT_VERSION@
 Requires:
-Libs: -L${libdir} -l@PROJECT_NAME_LOWER@
+Libs: -L${libdir} -l@PROJECT_NAME_LOWER@ -Wl,-rpath,${libdir}
 Cflags: -I${includedir}
```

However, it's true that rpath should not be set for standard directories (/lib, /usr/lib, ...)

But what can I do on the user side?

**Comment by MrAnno** [comment]
Adding this unconditionally would not be great, so I'm thinking of doing something conditionally or optionally.
Detecting whether we install Criterion to a system path is not really elegant in my opinion, because it's entirely up to the user what they set up as a library run path. In addition, RPATH is not completely portable, it behaves slightly differently on different platforms.

I also agree it wouldn't be a nice thing to leave this completely to the user.
I think we could define a pkg-config variable called `rpath`, which could optionally be used when the user of the library wants a hard-coded rpath, but their build system does not support this automatically (like an autotools+libtool combination does):

```
$ pkg-config --variable rpath criterion
-Wl,-rpath,...
```

Would that be okay with you?

(Meson uses its own generator for pkg-config, the mentioned `src/criterion.pc.in` file is just some leftover from the old CMake build. I'll remove that so as not to cause confusion.)

**Comment by Snaipe** [comment]
Alternatively I'd imagine we could add a meson option to insert the rpath flag for nonstandard prefixes by default. That way if a user decides to install to /usr/local, they can pass e.g. `-Dset_rpath=true` or something and the installed pkg-config file would append the -Wl,-rpath flag.

**Comment by MrAnno** [comment]
Resolved in #416.
It will be part of v2.4.0.
